### PR TITLE
add support for AMD AMF hardware encoding on Windows & Linux

### PR DESCRIPTION
**Pull request by nyanmisaka**
**Changes**
1) Add support for AMD AMF hardware encoding on Windows and Linux which can reduce cpu usage instead of using default libx264 in ffmpeg.
2) h264_amf is [now supported on linux](https://github.com/GPUOpen-LibrariesAndSDKs/AMF/issues/4#issuecomment-572301302) with ['amdgpu-pro'](https://amdgpu-install.readthedocs.io/en/latest/) installed and '--enable-amf' when compiling [ffmpeg](https://git.ffmpeg.org/gitweb/ffmpeg.git/commit/f8ad2ddd7a51df5d6e768ba2a6f65ee64b94e29b).
3) Using vaapi decode and h264_amf encode on linux platform can avoid some weird transcoding [errors in h264_vaapi with amd gpu](https://github.com/jellyfin/jellyfin/issues/1527).

It works like [this](https://imgur.com/Q3GrwkS) on windows now.
ffmpeg h264_amf on linux encoding succeed [logs](https://raw.githubusercontent.com/nyanmisaka/notepad/master/ffmpeg_h264_amf_on_linux_log?token=ADSCUQFAXHGMECH37SDK6EC6ERPD6) 

**Fixes**
issue #1420

### CODE REVIEW

**Comment by barolo** [comment]
It's worth to note that amf is also supported under Linux with closed drivers and soon enough should be with the open drivers too.

**Comment by nyanmisaka** [comment]
> It's worth to note that amf is also supported under Linux with closed drivers and soon enough should be with the open drivers too.



> It's worth to note that amf is also supported under Linux with closed drivers and soon enough should be with the open drivers too.

Sure. Since I often encounter strange errors when transcoding using the mesa open source driver VAAPI, I am currently trying to use AMF to replace it. Just learned that you need to install 'amf-amdgpu-pro' after installing the closed source driver. It seems that few people use AMF on Linux now.

**Comment by barolo** [comment]
> 
> Sure. Since I often encounter strange errors when transcoding using the mesa open source driver VAAPI, I am currently trying to use AMF to replace it. Just learned that you need to install 'amf-amdgpu-pro' after installing the closed source driver. It seems that few people use AMF on Linux now.

Yeah, vaapi for amd is not good plus apparently no one really cares about fixing it [ amf might be the reason ] I've opened related issue #1527 some time ago and decided that fixing it is not worth it because of the amf

**Comment by nyanmisaka** [comment]
> > Sure. Since I often encounter strange errors when transcoding using the mesa open source driver VAAPI, I am currently trying to use AMF to replace it. Just learned that you need to install 'amf-amdgpu-pro' after installing the closed source driver. It seems that few people use AMF on Linux now.
> 
> Yeah, vaapi for amd is not good plus apparently no one really cares about fixing it [ amf might be the reason ] I've opened related issue #1527 some time ago and decided that fixing it is not worth it because of the amf

I use h264_amf encoder in ffmpeg on manjaro to output video successfully with vaapi hardware acceleration. It should be noted that currently hevc_amf does not support Linux platforms. [ffmpeg logs here](https://raw.githubusercontent.com/nyanmisaka/notepad/master/h264_amf_succeed_on_linux.log?token=ADSCUQEVQY4PNCGZW4KBRQS6EQXKA)

**Comment by barolo** [comment]
@nyanmisaka does hevc 10bit transcode to h264 via amf without any issues? I can't test closed driver really with my distro.

**Comment by nyanmisaka** [comment]
> @nyanmisaka does hevc 10bit transcode to h264 via amf without any issues? I can't test closed driver really with my distro.

no issues anymore. 
transcoding from hevc@1080p(yuv420p10le) to h264@720p(yuv420p). [ffmpeg_log](https://raw.githubusercontent.com/nyanmisaka/notepad/master/h265p10_to_h264720p4m.log?token=ADSCUQAJ4OR5UOMAYDZSUC26ESSEU)
jellyfin can recognize amf encoder as well. [jellyfin_log](https://raw.githubusercontent.com/nyanmisaka/notepad/master/jellyfin_init_avaliable_codecs.log?token=ADSCUQETRME4JM6HW6RJWLK6ESS3Q)

**Comment by JustAMan** [comment]
@nyanmisaka can you add some text to our docs which explains how to enable AMF in the system? And probably an issue at `jellyfin-ffmpeg` repo to add `--enable-amf` if our ffmpeg doesn't support that yet.

File: MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs [hunk 1]
```diff
@@ -2642,22 +2642,11 @@ public string GetHardwareAcceleratedVideoDecoder(VideoType videoType, MediaStrea
...
-                            }
-                            break;
-                        case "mpeg2video":
-                            if (_mediaEncoder.SupportsDecoder("hevc_amf") && encodingOptions.HardwareDecodingCodecs.Contains("mpeg2video", StringComparer.OrdinalIgnoreCase))
-                            {
-                                return "-c:v mpeg2_mmal";
-                            }
-                            break;
+                        //h264_amf is now supported on linux with 'amdgpu-pro' installed and '--enable-amf' when compiling ffmpeg
+                        //using vaapi decode and h264_amf encode on linux platform can avoid some weird transcoding errors in h264_vaapi with amd gpu
```
Comment (JustAMan) [inline MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs]: This comment should not be there, don't treat code like a place to explain _why_ you did a change, comments should explain what the current (i.e. after your change) code does, not why did you change it. I think you should just drop this comment altogether.

File: MediaBrowser.MediaEncoding/Encoder/EncoderValidator.cs [hunk 2]
```diff
@@ -55,7 +55,9 @@ public class EncoderValidator
             "h264_vaapi",
             "hevc_vaapi",
             "h264_v4l2m2m",
-            "ac3"
+            "ac3",
+			"h264_amf",
+			"hevc_amf"
```
Comment (JustAMan) [inline MediaBrowser.MediaEncoding/Encoder/EncoderValidator.cs]: indentation seems off

**Comment by nyanmisaka** [comment]
> @nyanmisaka can you add some text to our docs which explains how to enable AMF in the system? And probably an issue at `jellyfin-ffmpeg` repo to add `--enable-amf` if our ffmpeg doesn't support that yet.

Once I have time I will add some AMD AMF instructions in the docs. It seems that only linux platform jellyfin-ffmpeg has not added `--enable-amf`.

File: MediaBrowser.MediaEncoding/Encoder/EncoderValidator.cs (hunk 2 above)
Comment (nyanmisaka) [reply MediaBrowser.MediaEncoding/Encoder/EncoderValidator.cs]: Thanks for reminding.

**Comment by goldzy123** [comment]
how to enable amf encoder in windows? using 20200228 nightly still shows 0% for video encode.

**Comment by nyanmisaka** [comment]
> how to enable amf encoder in windows? using 20200228 nightly still shows 0% for video encode.

1. Make sure you Radeon software is up to date.
2. Only higher than RX400 series can hw decode HEVC 10-bit video.
3. H264 10-bit hw decode is not supported.
4. The AMF encoder is test passed on my Vega 11 iGPU and RX Vega.
<details>
<summary>screenshot</summary>

vega11
</details>

Please post your ffmpeg logs and gpu model for further diagnosis.

**Comment by goldzy123** [comment]
> > how to enable amf encoder in windows? using 20200228 nightly still shows 0% for video encode.
> 
> 1. Make sure you Radeon software is up to date.
> 2. Only higher than RX400 series can hw decode HEVC 10-bit video.
> 3. H264 10-bit hw decode is not supported.
> 4. The AMF encoder is test passed on my Vega 11 iGPU and RX Vega.
> 
> screenshot
> Please post your ffmpeg logs and gpu model for further diagnosis.
amd 2200g cpu,winserver 2019.

[https://github.com/goldzy123/uptest/blob/master/123.png](url)
[https://github.com/goldzy123/uptest/blob/master/Log.txt](url)

**Comment by nyanmisaka** [comment]
> [https://github.com/goldzy123/uptest/blob/master/123.png](url)
> [https://github.com/goldzy123/uptest/blob/master/Log.txt](url)

`C:\Users\Administrator\Desktop\jellyfin_10.4.3\ffmpeg.exe`

Seems like you are using version 10.4.3? 

Try using the version [here ](https://repo.jellyfin.org/releases/server/windows/nightly/)without overwriting the older installation.

**Comment by goldzy123** [comment]
> > [https://github.com/goldzy123/uptest/blob/master/123.png](url)
> > [https://github.com/goldzy123/uptest/blob/master/Log.txt](url)
> 
> `C:\Users\Administrator\Desktop\jellyfin_10.4.3\ffmpeg.exe`
> 
> Seems like you are using version 10.4.3?
> AMF encoder in only supported in 10.5.0 nightly.
> 
> Try using the version [here ](https://repo.jellyfin.org/releases/server/windows/nightly/)without overwriting the older installation.

thank u for your help, i have updated to 10.5.0. Log shows :
[h264 @ 000001cdf285f040] No device available for decoder: device type d3d11va needed for codec h264.
Stream mapping:
  Stream #0:0 -> #0:0 (h264 (native) -> h264 (h264_amf))
  Stream #0:1 -> #0:1 (dts (dca) -> mp3 (libmp3lame))
Device setup failed for decoder on input stream #0:0 : Unknown error occurred

I think it's a driver issue now, amd's radeon driver doesnt support win server officially.
 thank u for your help again!

**Comment by KristupasSavickas** [comment]
Thanks for the feature! Has anyone had any luck setting this up on Linux? Docker?

**Comment by nyanmisaka** [comment]
> Thanks for the feature! Has anyone had any luck setting this up on Linux? Docker?
> 
> Found this PR in ffmpeg which might help someone: https://git.ffmpeg.org/gitweb/ffmpeg.git/commit/f8ad2ddd7a51df5d6e768ba2a6f65ee64b94e29b.

See this https://github.com/jellyfin/jellyfin/issues/2473

AMF does support both Windows and Linux.
As far as I know AMF currently only works on real machines.  Compared with NVIDIA, AMD has a long way to go in terms of video codec software, which requires time to catch up.

**Comment by janfribogar** [comment]
Hello @nyanmisaka can you help me? I'm having problems decoding with AMD AMF in windows.
Here you have my logs and specs:

Headless server runing Windows 10 Pro, AMD drivers 20.4.2 and Jellyfin 10.5.5
With the same setup but with Windows server 2019 it works perfectly (decoding and ecoding). Same driver version and same version of Jellyfin.

imagen

[ffmpeg-transcode-39793177-7f15-45db-9323-923a7779d5eb.txt](https://github.com/jellyfin/jellyfin/files/4634776/ffmpeg-transcode-39793177-7f15-45db-9323-923a7779d5eb.txt)
